linkerd/linkerd2
# Restrict the number of IPs an ExternalWorkload can have

A Kubernetes pod may be assigned at [most one IP address][pod-docs] for each supported protocol (i.e. IPv6 and IPv4), without the use of specialised CNIs or network configurations. When processing addresses in an endpoint, we will only ever use one address.

ExternalWorkload resources have a generic workloadIPs field that allow any number of addresses to be added. We want the behaviour to be similar to a pod -- only one address (of each protcol) should be used for routing.

We restrict the CRD server-side validation to allow only one IP address. Since we do not yet support IPv6, this will ensure that two IPv4 addresses will not be declared by the same workload. Once IPv6 support lands, or once we have a dedicated validator, we can relax the CRD validation.

In addition to the CRD change, we also refactor the logic around collecting IP addresses from EndpointSlice resources. The previous logic could mislead people into thinking we were processing _more than one IP address_ when in reality we were always overriding the ID (so even though we looked at more than one, we essentially always ended up with one in the set). We add two test cases to build confidence in the non-functional change.

[pod-docs]: https://pkg.go.dev/k8s.io/kubernetes@v1.29.1/pkg/apis/core#PodStatus

### How to test

* Install Linkerd after building the branch (just the crds will do, `linkerd install --crds`).
* Try to apply the following CRD:

```yaml
apiVersion: workload.linkerd.io/v1alpha1
kind: ExternalWorkload
metadata:
  labels:
    app: legacy
  name: external-workload-invalid
  namespace: mixed-env
spec:
  meshTls:
    identity: spiffe://root.linkerd.cluster.local/external-workload-invalid
    serverName: external-workload-invalid.cluster.local
  ports:
  - name: http
    port: 80
    protocol: TCP
  workloadIPs:
  - ip: 172.22.0.5
  - ip: 172.22.0.6
status:
  conditions:
  - lastTransitionTime: "2024-01-24T11:53:43Z"
    message: This workload is alive
    reason: Alive
    status: "True"
    type: Ready
```

* Expect the creation to fail

> The ExternalWorkload "external-workload-invalid" is invalid: spec.workloadIPs: Too many: 2: must have at most 1 items
